src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs

```diff
@@ -0,0 +1,151 @@
+using Xunit;
+
+namespace Stratis.Bitcoin.Tests.Consensus
+{
+    public class ChainedHeaderTreeTest
+    {
+        public class TestContext
+        {
+            public Network Network = Network.RegTest;
+            public Mock<IChainedHeaderValidator> ChainedHeaderValidatorMock = new Mock<IChainedHeaderValidator>();
+            public Mock<ICheckpoints> CheckpointsMock = new Mock<ICheckpoints>();
+            public Mock<IChainState> ChainStateMock = new Mock<IChainState>();
+            public ConsensusSettings ConsensusSettings = new ConsensusSettings();
+
+            public ChainedHeaderTree ChainedHeaderTree;
+
+            public ChainedHeaderTree CreateChainedHeaderTree()
+            {
+                this.ChainedHeaderTree = new ChainedHeaderTree(this.Network, new ExtendedLoggerFactory(), this.ChainedHeaderValidatorMock.Object, this.CheckpointsMock.Object, this.ChainStateMock.Object, this.ConsensusSettings);
+                return this.ChainedHeaderTree;
+            }
+
+            public ChainedHeader ExtendAChain(int count, ChainedHeader chainedHeader = null)
+            {
+                ChainedHeader previousHeader = chainedHeader ?? new ChainedHeader(this.Network.GetGenesis().Header, this.Network.GenesisHash, 0);
+
+                foreach (int index in Enumerable.Range(1, count))
+                {
+                    BlockHeader header  = this.Network.Consensus.ConsensusFactory.CreateBlockHeader();
+                    header.HashPrevBlock = previousHeader.HashBlock;
+                    header.Bits = previousHeader.Header.Bits - 1000; // just increase difficulty.
+                    ChainedHeader newHeader = new ChainedHeader(header, header.GetHash(), previousHeader);
+                    previousHeader = newHeader;
+                }
+
+                return previousHeader;
+            }
+
+            public List<BlockHeader> ChainedHeaderToList(ChainedHeader chainedHeader, int count)
+            {
+                List<BlockHeader> list = new List<BlockHeader>();
+
+                ChainedHeader current = chainedHeader;
+                while (count > 0)
+                {
+                    list.Add(current.Header);
+                    current = current.Previous;
+                    count--;
+                }
+
+                list.Reverse();
+
+                return list;
+            }
+
+            public bool ConnectedHeadersIsEmpty(ConnectedHeaders connectedHeaders)
+            {
+                Assert.NotNull(connectedHeaders);
+
+                return connectedHeaders.Consumed == null 
+                       && connectedHeaders.DownloadTo == null 
+                       && connectedHeaders.DownloadFrom == null;
+            }
+        }
+
+        [Fact]
+        public void ConnectHeaders_HeadersCantConnect_ShouldFail()
+        {
+            TestContext testContext = new TestContext();
+            ChainedHeaderTree chainedHeaderTree = testContext.CreateChainedHeaderTree();
```

Comment:
How is the TestContext class better than just having a private method in this class for things like `CreateChainedHeaderTree()`  and the state variables within the class directly too?

Reply:
The usefulness of the context is to allow changing the mocks per each test.

Also there will be a following PR that will focus on tests, so any test improvements we could probably do there.